src/test/epoll/test_epoll.rs

```diff
@@ -116,6 +124,112 @@ fn test_threads_level() -> anyhow::Result<()> {
     Ok(())
 }
 
+/// This test has the threads read from the pipe immediately after returning from `epoll_wait`.
+///
+/// In Linux, if the first thread to wake from `epoll_wait` reads from the pipe fast enough, the
+/// other threads blocked on `epoll_wait` won't wake up. This is a race condition, but it means that
+/// not all threads will necessarily be woken (unlike the `test_threads_level` test above). Since
+/// Shadow doesn't switch threads until there is a blocking syscall, we expect that the thread which
+/// first wakes from `epoll_wait` will read from the pipe before the other thread is woken from
+/// `epoll_wait`, which means the second thread shouldn't be woken until the timeout expires.
+///
+/// The purpose of this test is to make sure Shadow replicates the Linux behaviour of sometimes
+/// avoiding a spurious wakeup if the epoll status changes back to "not readable" before the thread
+/// has a chance to run.  Code that uses epoll shouldn't rely on all `epoll_wait` calls being woken.
+fn test_threads_level_with_late_read() -> anyhow::Result<()> {
+    let (readfd, writefd) = unistd::pipe2(nix::fcntl::OFlag::O_NONBLOCK)?;
+    let epollfd = epoll::epoll_create()?;
+
+    let mut event = epoll::EpollEvent::new(EpollFlags::EPOLLIN, readfd as u64);
+    epoll::epoll_ctl(
+        epollfd,
+        epoll::EpollOp::EpollCtlAdd,
+        readfd,
+        Some(&mut event),
+    )?;
+
+    let timeout = Duration::from_millis(100);
+
+    let threads = [
+        std::thread::spawn(move || do_epoll_wait(epollfd, timeout, /* do_read= */ true)),
+        std::thread::spawn(move || do_epoll_wait(epollfd, timeout, /* do_read= */ true)),
+    ];
+
+    // Wait for readers to block.
+    std::thread::sleep(timeout / 2);
+
+    // Make the read-end readable.
+    unistd::write(writefd, &[0])?;
+
+    let mut results = threads.map(|t| t.join().unwrap());
+
+    // One of the threads should have gotten an event, but we don't know which one.
+    // Sort results by number of events received.
+    results.sort_by(|lhs, rhs| lhs.events.len().cmp(&rhs.events.len()));
+
+    // One thread should have timed out with no events received.
+    ensure_ord!(results[0].epoll_res, ==, Ok(0));
+    ensure_ord!(results[0].duration, >=, timeout);
```

Comment:
I'm confused how this passes without https://github.com/shadow/shadow/pull/2759 . It'd be nice if there's a small tweak or variation we could do to these tests to act as a regression test for that bugfix.

Reply:
I think it works because in this case the process is restarted early, the syscall condition is cancelled, and the `epoll_wait` syscall handler runs again, but sees that it should re-block. A new syscall condition is then created with a new timer. I'm not sure what happens to the old timer event though, maybe there's now two of them in the event queue?

I'm not sure of how we could write a regression test for #2759 though.